src/dir/glob.cr

```diff
@@ -1,266 +1,369 @@
+            alternatives << pattern.byte_slice(start, reader.pos - start)
+            break
+          end
+        when ','
+          if nest == 1
+            start = (alt_start || lbrace).not_nil! + 1
+            alternatives << pattern.byte_slice(start, reader.pos - start)
+            alt_start = reader.pos
+          end
         end
+
+        reader.next_char
       end
-    end
-  end
 
-  private def self.glob2regex(pattern)
-    if pattern.size == 0 || pattern == File::SEPARATOR
-      raise ArgumentError.new "Empty glob pattern"
-    end
+      if lbrace && rbrace
+        front = pattern.byte_slice(0, lbrace)
+        back = pattern.byte_slice(rbrace + 1)
 
-    # characters which are escapable by a backslash in a glob pattern;
-    # Windows paths must have double backslashes:
-    escapable = {'?', '{', '}', '*', ',', '\\'}
-    # characters which must be escaped in a PCRE regex:
-    escaped = {'$', '(', ')', '+', '.', '[', '^', '|', '/'}
+        alternatives.each do |alt|
+          brace_pattern = [front, alt, back].join
 
-    last_is_file_separator = false
+          compile brace_pattern, patterns
+        end
+      else
+        patterns << single_compile pattern
+      end
 
-    regex_pattern = String.build do |str|
-      str << "\\A"
-      idx = 0
-      nest = 0
+      patterns
+    end
 
-      idx = 1 if pattern[0] == File::SEPARATOR
-      size = pattern.size
+    private def single_compile(glob)
+      list = [] of PatternType
+      return list if glob.empty?
 
-      while idx < size
-        char = pattern[idx]
+      parts = glob.split('/', remove_empty: true)
 
-        if last_is_file_separator && char == File::SEPARATOR
-          idx += 1
-          next
+      if glob.ends_with?('/')
+        list << DirectoriesOnly.new
+      else
+        file = parts.pop
+        if /^[a-zA-Z0-9._]+$/.match(file)
+          list << ConstantEntry.new file
+        elsif !file.empty?
+          list << EntryMatch.new file
         end
+      end
 
-        last_is_file_separator = char == File::SEPARATOR
+      while !parts.empty?
+        dir = parts.pop
 
-        if char == '\\'
-          if idx + 1 < size && escapable.includes?(peek = pattern[idx + 1])
-            str << '\\'
-            str << peek
-            idx += 2
-            next
+        case dir
+        when "**"
+          if parts.empty?
+            list << StartRecursiveDirectories.new
+          else
+            list << RecursiveDirectories.new
           end
-        elsif char == '*'
-          if idx + 2 < size &&
-             pattern[idx + 1] == '*' &&
-             pattern[idx + 2] == File::SEPARATOR
-            str << "(?:.*\\" << File::SEPARATOR << ")?"
-            idx += 3
-            next
-          elsif idx + 1 < pattern.size && pattern[idx + 1] == '*'
-            str << "[^\\" << File::SEPARATOR << "]*"
-            idx += 2
-            next
+        when /^[^\*\?\]]+$/
```

Comment:
I wonder if there's a way to avoid regexes at all.

Still, your changes are an improvement over the original code, so it doesn't matter much (it's better if it works well first)

Reply:
Yes, this can be improved. Though I think it is okay to use regexes for these tasks, since compile is only run once per glob run, not for every matching file.

Reply:
No problem, performance can always be improved later :-)